### PR TITLE
[POAE7-2812] Left Join Implement

### PR DESCRIPTION
**Pull request by qiuyangshen**
### What changes were proposed in this pull request?

- Left Join function implement
- delete some unnecessary header files
- add related left join test
- TODO: will extract buildTableReader as a utility class and  merged with ColumnToRowReader in next PR

### Why are the changes needed?
For NextGen HashJoin op funciton

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UTs

### Which label does this PR belong to?
INFRA

### CODE REVIEW

**Comment by jikunshang** [comment]
Any design doc, workflow diagrams for join codegen part? I think it's hard to understand/review/maintain without it.

**Comment by qiuyangshen** [comment]
> Any design doc, workflow diagrams for join codegen part? I think it's hard to understand/review/maintain without it.

Get it

File: cpp/src/cider/exec/nextgen/operators/HashJoinNode.cpp [hunk 1]
```diff
@@ -169,6 +234,67 @@ void traverse(ExprPtr expr,
...
+          .params_vector = {join_res_buffer.get(), row_index.get()}});
+
+  auto res_row_id = func->emitRuntimeFunctionCall(
+      "extract_join_row_id",
+      JITFunctionEmitDescriptor{
+          .ret_type = JITTypeTag::INT64,
+          .params_vector = {join_res_buffer.get(), row_index.get()}});
+
+  std::map<ExprPtr, size_t>::iterator iter;
+  for (iter = build_table_map.begin(); iter != build_table_map.end(); iter++) {
```
[the remark this answers is not in the document]
Comment (qiuyangshen) [reply cpp/src/cider/exec/nextgen/operators/HashJoinNode.cpp:260]: All right. `build_table_map` is used to preserve the desired build table expr and offset, such as selected project expr and join key when traverse EU.  then L282&L283 will read the corresponding column value.

[the remark this answers is not in the document]
Comment (jikunshang) [reply cpp/src/cider/exec/nextgen/operators/HashJoinNode.cpp:260]: `SELECT t_l.col_a, t_r.col_a FROM t_l JOIN t_r ON t_l.col_a = t_r.col_a`, this case `build_table_map` will contain `t_r.col_a` ?
`SELECT t_l.col_a, t_r.col_b FROM t_l JOIN t_r ON t_l.col_a = t_r.col_a`, this case `build_table_map` will contain `t_r.col_a` and `t_r.col_b`  ? and will read `t_r.col_a` even we don't actually need it after probe?

[the remark this answers is not in the document]
Comment (qiuyangshen) [reply cpp/src/cider/exec/nextgen/operators/HashJoinNode.cpp:260]: Yes, will get the selected column when traversing eu.shared_target_exprs, and get the join key column by the eu.join_quals

**Review by xieqi — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Pull request merged**